Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,420 @@
+
+  rw [intervalIntegral.integral_congr this, intervalIntegral.integral_const, smul_eq_mul]
+
+/--
+  Buffon's Needle, the short case (`d ≤ l`).
+-/
+theorem buffon_long (h : d ≤ l) :
+    ℙ[N l B] = (2 * l) / (d * π) - 2 / (d * π) * ((l^2 - d^2).sqrt + d * (d / l).arcsin) + 1 := by
+
+  simp only [
+    buffon_integral d l hd B hBₘ hB, MeasureTheory.integral_const, smul_eq_mul, mul_one,
+    MeasurableSet.univ, Measure.restrict_apply, Set.univ_inter, Set.Icc_inter_Icc, Real.volume_Icc,
+    sup_eq_max, inf_eq_min, min_div_div_right zero_le_two d, max_div_div_right zero_le_two (-d),
+    div_sub_div_same, neg_mul, max_neg_neg, sub_neg_eq_add, ← mul_two,
+    mul_div_cancel (min d (Real.sin _ * l)) two_ne_zero
+  ]
+
+  have (θ : ℝ) (hθ : θ ∈ Set.Icc 0 π) : min d (θ.sin * l) ≥ 0 := by
+    by_cases h : d ≤ θ.sin * l
+    · rw [min_eq_left h]; exact hd.le
+    · rw [min_eq_right (not_le.mp h).le]; exact mul_nonneg (Real.sin_nonneg_of_mem_Icc hθ) hl.le
+
+  rw [set_integral_toReal_ofReal_nonneg_ae measurableSet_Icc (MeasureTheory.ae_of_all _ this),
+    MeasureTheory.integral_Icc_eq_integral_Ioc,
+    ← intervalIntegral.integral_of_le Real.pi_pos.le, integral_min_eq_two_mul,
+    ← intervalIntegral.integral_add_adjacent_intervals
+      (min_const_sin_mul_intervalIntegrable d l _ _) (min_const_sin_mul_intervalIntegrable d l _ _),
+    integral_zero_to_arcsin_min d l hd hl, integral_arcsin_to_pi_div_two_min d l hl h]
+
+  /-
+    the rest of the proof is annoying algebra, the goal is
+      (d * π)⁻¹ * (2 * ((1 - sqrt (1 - (d / l) ^ 2)) * l + (π / 2 - arcsin (d / l)) * d)) =
+      2 * l / (d * π) - 2 / (d * π) * (sqrt (l ^ 2 - d ^ 2) + d * arcsin (d / l)) + 1
+  -/
+
+  have this₁ : (1 - Real.sqrt (1 - (d / l) ^ 2)) * l = l - (l ^ 2 - d ^ 2).sqrt := by
+    rw [mul_comm, mul_sub, mul_one, div_pow, one_sub_div, Real.sqrt_div, Real.sqrt_sq hl.le,
+      ← mul_div_assoc, mul_comm, mul_div_cancel _ (ne_of_gt hl)]
+    · simp_rw [sub_nonneg, sq_le_sq, abs_of_pos hd, abs_of_pos hl, h]
+    · simp_rw [(pow_eq_zero_iff two_ne_zero).not, ne_of_gt hl, not_false_eq_true]
+  have this₂ : 2 * d * (π / 2 - (d / l).arcsin) / (d * π) = 1 - (2 / π) * (d / l).arcsin := by
+    rw [mul_sub, sub_div, mul_assoc, ← mul_comm_div, ← mul_assoc, ← mul_comm_div,
+      div_self two_ne_zero, one_mul, div_self (ne_of_gt (mul_pos hd Real.pi_pos)), mul_div_assoc,
+      ← mul_comm_div, mul_comm 2, mul_div_mul_left _ _ (ne_of_gt hd)]
+  have this₃ : 2 * Real.sqrt (l^2 - d^2) / (d * π) = 2 / (d * π) * (l^2 - d^2).sqrt := by ring_nf
+  have this₄ : 2 / π * d / d = 2 / (d * π) * d := by ring_nf
+
+  conv =>
+    lhs
+    rw [this₁, inv_mul_eq_div, mul_add, mul_sub, add_div, sub_div,
+      mul_comm (π / 2 - (d / l).arcsin), ← mul_assoc, this₂, this₃, add_sub, add_sub_right_comm,
+      sub_eq_add_neg, sub_eq_add_neg, ← neg_mul, ← mul_div_cancel (2 / π) (ne_of_gt hd), this₄,
+      mul_assoc, ← neg_mul, add_assoc (2 * l / (d * π)) _ _, ← mul_add, neg_mul, ← sub_eq_add_neg]
```

Comment:
Fortunately computers are better at this:
```suggestion
  field_simp
  ring
```

Reply:
Wow. I wasn't aware of `field_simp`. One question: Usually, when I use `ring`, Lean will ask me instead to replace it with `ring_nf`. Here, it doesn't say that. Do you know why?

Reply:
`ring` has several subroutines. If the first one doesn't work, it tells you to run `ring_nf`, which is the second subroutine. `_nf` stands for "Horner normed form".